msticpy/nbtools/timeline.py

```diff
@@ -0,0 +1,940 @@
+
+
+# pylint: enable=too-many-locals
+
+
+def _wrap_df_columns(data: pd.DataFrame, wrap_len: int = 50):
+    """Wrap any string columns."""
+    for col in data.columns:
+        if isinstance(data[col].iloc[0], str):
+            data[col] = data[col].str.wrap(wrap_len)
+
+
+def _get_tick_formatter() -> DatetimeTickFormatter:
+    """Return tick formatting for different zoom levels."""
+    # '%H:%M:%S.%3Nms
+    tick_format = DatetimeTickFormatter()
+    tick_format.days = ["%m-%d %H:%M"]
+    tick_format.hours = ["%H:%M:%S"]
+    tick_format.minutes = ["%H:%M:%S"]
+    tick_format.seconds = ["%H:%M:%S"]
+    tick_format.milliseconds = ["%H:%M:%S.%3N"]
+    return tick_format
+
+
+def _calc_auto_plot_height(group_count):
+    """Dynamic calculation of plot height."""
+    ht_per_row = 40
+    if group_count > 15:
+        ht_per_row = 25
+    return max(ht_per_row * group_count, 300)
+
+
+# pylint: disable=too-many-arguments
+def _create_range_tool(
+    data, min_time, max_time, plot_range, width, time_column: str = None
+):
+    """Create plot bar to act as as range selector."""
+    ext_min = min_time - ((max_time - min_time) * 0.15)
+    ext_max = max_time + ((max_time - min_time) * 0.15)
+    rng_select = figure(
+        x_range=(ext_min, ext_max),
+        title="Range Selector",
+        plot_height=100,
+        plot_width=width,
+        x_axis_type="datetime",
+        y_axis_type=None,
+        tools="",
+        toolbar_location=None,
+    )
+    help_str = (
+        "Drag the middle or edges of the selection box to change "
+        + "the range in the main chart"
+    )
+    rng_select.add_layout(
+        Title(text=help_str, align="right", text_font_size="10px"), "below"
+    )
+    rng_select.xaxis[0].formatter = _get_tick_formatter()
+    if isinstance(data, dict):
+        for _, series_def in data.items():
+            rng_select.circle(
+                x=series_def["time_column"],
+                y="y_index",
+                color=series_def["color"],
+                source=series_def["source"],
+            )
+    elif isinstance(data, pd.DataFrame):
+        rng_select.circle(
+            x=time_column, y="y_index", color="blue", source=ColumnDataSource(data)
+        )
+
+    range_tool = RangeTool(x_range=plot_range)
+    range_tool.overlay.fill_color = "navy"
+    range_tool.overlay.fill_alpha = 0.2
+    rng_select.ygrid.grid_line_color = None
+    rng_select.add_tools(range_tool)
+    rng_select.toolbar.active_multi = range_tool
+    return rng_select
+
+
+# pylint: enable=too-many-arguments
+
+
+def _add_ref_line(plot, ref_time, series_count=1):
+    """Add a reference marker line and label at `ref_time`."""
+    ref_label_tm = pd.Timestamp(ref_time)
+    plot.line(x=[ref_label_tm, ref_label_tm], y=[0, series_count])
+    ref_label = Label(
+        x=ref_label_tm,
+        y=0,
+        y_offset=10,
+        x_units="data",
+        y_units="data",
+        text="< ref",
```

Comment:
This text isn't particularly useful when using the timeline, we should have it as a kwarg we can pass with the ref object or if we pass it an alert at least set this to 'Alert'

Reply:
implement reference label param (infer if alert supplied)